### PR TITLE
nvmftest: Using ssh and scp utils

### PR DESCRIPTION
**Pull request by narasimhan-v**
Using the ssh and scp utils for nvmf tests.

Signed-off-by: Narasimhan V <sim@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by narasimhan-v** [comment]
```
# avocado run nvmftest.py:NVMfTest.test_targetconfig nvmftest.py:NVMfTest.test_cleartargetconfig -m nvmftest.py.data/nvmftest.yaml 
JOB ID     : afb47624a01f80365eee13a4fc9b1ace8f7ae39f
JOB LOG    : /root/tests/results/job-2020-09-27T17.05-afb4762/job.log
 (1/2) nvmftest.py:NVMfTest.test_targetconfig;run-3a9b: PASS (1.61 s)
 (2/2) nvmftest.py:NVMfTest.test_cleartargetconfig;run-3a9b: PASS (1.57 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/tests/results/job-2020-09-27T17.05-afb4762/results.html
JOB TIME   : 4.00 s
```

```
2020-09-27 17:05:20,437 test             L0530 INFO | START 1-nvmftest.py:NVMfTest.test_targetconfig;run-3a9b
2020-09-27 17:05:20,437 test             L0206 DEBUG| DATA (filename=output.expected) => NOT FOUND (data sources: variant, test, file)
2020-09-27 17:05:20,539 sysinfo          L0113 DEBUG| Not logging /proc/pci (file does not exist)
2020-09-27 17:05:20,551 sysinfo          L0113 DEBUG| Not logging /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor (file does not exist)
2020-09-27 17:05:20,668 parameters       L0143 DEBUG| PARAMS (key=namespaces, path=*, default=) => '/dev/nvme0n11 /dev/nvme0n12'
2020-09-27 17:05:20,668 parameters       L0143 DEBUG| PARAMS (key=peer_ips, path=*, default=) => '156.1.78.210 157.1.78.210'
2020-09-27 17:05:20,669 parameters       L0143 DEBUG| PARAMS (key=peer_user, path=*, default=root) => 'root'
2020-09-27 17:05:20,669 parameters       L0143 DEBUG| PARAMS (key=peer_password, path=*, default=None) => 'passw0rd'
2020-09-27 17:05:21,356 yum              L0068 DEBUG| dnf version: 4.2.17
2020-09-27 17:05:21,357 yum              L0050 ERROR| dnf module for Python is required. Using the basic support from rpm and dnf commands
2020-09-27 17:05:21,357 process          L0604 INFO | Running 'rpm -q nvme-cli'
2020-09-27 17:05:21,367 process          L0416 DEBUG| [stdout] nvme-cli-1.9-5.el8.ppc64le
2020-09-27 17:05:21,368 process          L0686 INFO | Command 'rpm -q nvme-cli' finished with 0 after 0.006455421447753906s
2020-09-27 17:05:21,403 test             L0202 DEBUG| DATA (filename=nvmf_template.cfg) => /root/avocado-misc-tests/io/nvmf/nvmftest.py.data/nvmf_template.cfg (found at file source dir)
2020-09-27 17:05:21,404 process          L0604 INFO | Running '/usr/bin/ssh -o 'StrictHostKeyChecking=no' -o 'UpdateHostKeys=no' -o 'ControlPath=~/.ssh/avocado-master-%r@%h:%p' -l root -O check 156.1.78.210 '''
2020-09-27 17:05:21,410 process          L0416 DEBUG| [stderr] Master running (pid=7557)
2020-09-27 17:05:21,410 process          L0686 INFO | Command '/usr/bin/ssh -o 'StrictHostKeyChecking=no' -o 'UpdateHostKeys=no' -o 'ControlPath=~/.ssh/avocado-master-%r@%h:%p' -l root -O check 156.1.78.210 ''' finished with 0 after 0.00520014762878418s
2020-09-27 17:05:21,436 process          L0604 INFO | Running '/usr/bin/ssh -o 'StrictHostKeyChecking=no' -o 'UpdateHostKeys=no' -o 'ControlPath=~/.ssh/avocado-master-%r@%h:%p' -l root -O check 156.1.78.210 '''
2020-09-27 17:05:21,443 process          L0416 DEBUG| [stderr] Master running (pid=7557)
2020-09-27 17:05:21,443 process          L0686 INFO | Command '/usr/bin/ssh -o 'StrictHostKeyChecking=no' -o 'UpdateHostKeys=no' -o 'ControlPath=~/.ssh/avocado-master-%r@%h:%p' -l root -O check 156.1.78.210 ''' finished with 0 after 0.005118608474731445s
2020-09-27 17:05:21,477 output           L0696 DEBUG| [stdout] /root/avocado-misc-tests/io/nvmf/nvmftest.py.data/nvmf.cfg
2020-09-27 17:05:21,477 output           L0696 DEBUG| [stdout]

2020-09-27 17:05:21,479 process          L0604 INFO | Running '/usr/bin/scp -o 'StrictHostKeyChecking=no' -o 'UpdateHostKeys=no' -o 'ControlPath=~/.ssh/avocado-master-%r@%h:%p' /root/avocado-misc-tests/io/nvmf/nvmftest.py.data/nvmf.cfg 156.1.78.210:/tmp/'
2020-09-27 17:05:21,506 process          L0686 INFO | Command '/usr/bin/scp -o 'StrictHostKeyChecking=no' -o 'UpdateHostKeys=no' -o 'ControlPath=~/.ssh/avocado-master-%r@%h:%p' /root/avocado-misc-tests/io/nvmf/nvmftest.py.data/nvmf.cfg 156.1.78.210:/tmp/' finished with 0 after 0.026220083236694336s
2020-09-27 17:05:21,562 process          L0604 INFO | Running '/usr/bin/ssh -o 'StrictHostKeyChecking=no' -o 'UpdateHostKeys=no' -o 'ControlPath=~/.ssh/avocado-master-%r@%h:%p' -l root -q 156.1.78.210 'modprobe nvmet''
2020-09-27 17:05:21,581 process          L0686 INFO | Command '/usr/bin/ssh -o 'StrictHostKeyChecking=no' -o 'UpdateHostKeys=no' -o 'ControlPath=~/.ssh/avocado-master-%r@%h:%p' -l root -q 156.1.78.210 'modprobe nvmet'' finished with 0 after 0.017683029174804688s
2020-09-27 17:05:21,582 process          L0604 INFO | Running '/usr/bin/ssh -o 'StrictHostKeyChecking=no' -o 'UpdateHostKeys=no' -o 'ControlPath=~/.ssh/avocado-master-%r@%h:%p' -l root -q 156.1.78.210 'modprobe nvmet-rdma''
2020-09-27 17:05:21,601 process          L0686 INFO | Command '/usr/bin/ssh -o 'StrictHostKeyChecking=no' -o 'UpdateHostKeys=no' -o 'ControlPath=~/.ssh/avocado-master-%r@%h:%p' -l root -q 156.1.78.210 'modprobe nvmet-rdma'' finished with 0 after 0.013961076736450195s
2020-09-27 17:05:21,601 process          L0604 INFO | Running '/usr/bin/ssh -o 'StrictHostKeyChecking=no' -o 'UpdateHostKeys=no' -o 'ControlPath=~/.ssh/avocado-master-%r@%h:%p' -l root -q 156.1.78.210 'which nvmetcli''
2020-09-27 17:05:21,619 process          L0416 DEBUG| [stdout] /usr/sbin/nvmetcli
2020-09-27 17:05:21,620 process          L0686 INFO | Command '/usr/bin/ssh -o 'StrictHostKeyChecking=no' -o 'UpdateHostKeys=no' -o 'ControlPath=~/.ssh/avocado-master-%r@%h:%p' -l root -q 156.1.78.210 'which nvmetcli'' finished with 0 after 0.017000198364257812s
2020-09-27 17:05:21,645 process          L0604 INFO | Running '/usr/bin/ssh -o 'StrictHostKeyChecking=no' -o 'UpdateHostKeys=no' -o 'ControlPath=~/.ssh/avocado-master-%r@%h:%p' -l root -q 156.1.78.210 'nvmetcli restore /tmp/nvmf.cfg''
2020-09-27 17:05:21,790 process          L0686 INFO | Command '/usr/bin/ssh -o 'StrictHostKeyChecking=no' -o 'UpdateHostKeys=no' -o 'ControlPath=~/.ssh/avocado-master-%r@%h:%p' -l root -q 156.1.78.210 'nvmetcli restore /tmp/nvmf.cfg'' finished with 0 after 0.14341306686401367s
```

**Comment by narasimhan-v** [comment]
Without sleep
```
2020-09-27 17:11:22,011 process          L0604 INFO | Running 'nvme list'
2020-09-27 17:11:22,014 process          L0686 INFO | Command 'nvme list' finished with 0 after 0.00131988525390625s
2020-09-27 17:11:22,014 output           L0696 DEBUG| [stdout] Count: 0
2020-09-27 17:11:22,014 output           L0696 DEBUG| [stdout]

2020-09-27 17:11:22,015 process          L0604 INFO | Running 'nvme connect -t rdma -n mysubsys1 -a 156.1.78.210 -s 4420'
2020-09-27 17:11:22,059 process          L0686 INFO | Command 'nvme connect -t rdma -n mysubsys1 -a 156.1.78.210 -s 4420' finished with 0 after 0.04232001304626465s
2020-09-27 17:11:22,086 process          L0604 INFO | Running 'nvme connect -t rdma -n mysubsys2 -a 157.1.78.210 -s 4420'
2020-09-27 17:11:22,131 process          L0686 INFO | Command 'nvme connect -t rdma -n mysubsys2 -a 157.1.78.210 -s 4420' finished with 0 after 0.0440065860748291s
2020-09-27 17:11:22,131 process          L0604 INFO | Running 'nvme list'
2020-09-27 17:11:22,135 process          L0416 DEBUG| [stdout] Node             SN                   Model                                    Namespace Usage                      Format           FW Rev
2020-09-27 17:11:22,135 process          L0416 DEBUG| [stdout] ---------------- -------------------- ---------------------------------------- --------- -------------------------- ---------------- --------
2020-09-27 17:11:22,135 process          L0416 DEBUG| [stdout] /dev/nvme0n1     722bd5720ee7d82b     Linux                                    1          30.01  GB /  30.01  GB      4 KiB +  0 B   4.18.0-1
2020-09-27 17:11:22,136 process          L0686 INFO | Command 'nvme list' finished with 0 after 0.001840353012084961s
2020-09-27 17:11:22,161 output           L0696 DEBUG| [stdout] Count: 1
2020-09-27 17:11:22,161 output           L0696 DEBUG| [stdout]
```
With sleep
```
2020-09-27 17:14:38,204 output           L0696 DEBUG| [stdout] Count: 0
2020-09-27 17:14:38,204 output           L0696 DEBUG| [stdout]

2020-09-27 17:14:38,205 process          L0604 INFO | Running 'nvme connect -t rdma -n mysubsys1 -a 156.1.78.210 -s 4420'
2020-09-27 17:14:38,248 process          L0686 INFO | Command 'nvme connect -t rdma -n mysubsys1 -a 156.1.78.210 -s 4420' finished with 0 after 0.04195356369018555s
2020-09-27 17:14:39,277 process          L0604 INFO | Running 'nvme connect -t rdma -n mysubsys2 -a 157.1.78.210 -s 4420'
2020-09-27 17:14:39,321 process          L0686 INFO | Command 'nvme connect -t rdma -n mysubsys2 -a 157.1.78.210 -s 4420' finished with 0 after 0.04184842109680176s
2020-09-27 17:14:40,324 process          L0604 INFO | Running 'nvme list'
2020-09-27 17:14:40,327 process          L0416 DEBUG| [stdout] Node             SN                   Model                                    Namespace Usage                      Format           FW Rev
2020-09-27 17:14:40,327 process          L0686 INFO | Command 'nvme list' finished with 0 after 0.0008609294891357422s
2020-09-27 17:14:40,327 process          L0416 DEBUG| [stdout] ---------------- -------------------- ---------------------------------------- --------- -------------------------- ---------------- --------
2020-09-27 17:14:40,328 process          L0416 DEBUG| [stdout] /dev/nvme0n1     db7d04c56c748d71     Linux                                    1          30.01  GB /  30.01  GB      4 KiB +  0 B   4.18.0-1
2020-09-27 17:14:40,328 process          L0416 DEBUG| [stdout] /dev/nvme1n1     4995a8a000d10806     Linux                                    2          30.01  GB /  30.01  GB      4 KiB +  0 B   4.18.0-1
2020-09-27 17:14:40,375 output           L0696 DEBUG| [stdout] Count: 2
2020-09-27 17:14:40,375 output           L0696 DEBUG| [stdout]
```

```
# avocado run nvmftest.py -m nvmftest.py.data/nvmftest.yaml 
JOB ID     : e3789f0aaeb56d28901bfee817e0fd65de038fb1
JOB LOG    : /root/tests/results/job-2020-09-27T17.14-e3789f0/job.log
 (1/7) nvmftest.py:NVMfTest.test_targetconfig;run-3a9b: PASS (1.60 s)
 (2/7) nvmftest.py:NVMfTest.test_nvmfdiscover;run-3a9b: PASS (1.39 s)
 (3/7) nvmftest.py:NVMfTest.test_nvmfconnect;run-3a9b: PASS (3.63 s)
 (4/7) nvmftest.py:NVMfTest.test_nvmfdisconnect;run-3a9b: PASS (1.75 s)
 (5/7) nvmftest.py:NVMfTest.test_nvmfconnectcfg;run-3a9b: PASS (1.63 s)
 (6/7) nvmftest.py:NVMfTest.test_nvmfdisconnectcfg;run-3a9b: PASS (1.74 s)
 (7/7) nvmftest.py:NVMfTest.test_cleartargetconfig;run-3a9b: PASS (1.55 s)
RESULTS    : PASS 7 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/tests/results/job-2020-09-27T17.14-e3789f0/results.html
JOB TIME   : 14.38 s
```

**Review by smruti77 — APPROVED** [review]

**Comment by narasimhan-v** [comment]
@harish-24 please merge, if no further review comments

File: io/nvmf/nvmftest.py [hunk 1]
```diff
@@ -96,28 +101,34 @@ def nvme_devs_count():
...
-            if process.system(cmd, shell=True, ignore_status=True) != 0:
+            cmd = "modprobe %s" % mdl
+            output = self.session.cmd(cmd)
+            if not output.exit_status == 0:
                 self.cancel("%s is not loadable on the peer" % mdl)
         msg = "which nvmetcli"
-        cmd = "ssh %s \"%s\"" % (self.peer_ips[0], msg)
-        if process.system(cmd, shell=True, ignore_status=True) != 0:
+        output = self.session.cmd(msg)
+        if not output.exit_status == 0:
```
Comment (harish-24) [inline io/nvmf/nvmftest.py:127]: `if output.exit_status:`

File: io/nvmf/nvmftest.py [hunk 2]
```diff
@@ -96,28 +101,34 @@ def nvme_devs_count():
...
             self.cancel("nvmetcli is not installed on the peer")
         msg = "nvmetcli restore /tmp/nvmf.cfg"
-        cmd = "ssh %s \"%s\"" % (self.peer_ips[0], msg)
-        if process.system(cmd, shell=True, ignore_status=True) != 0:
+        output = self.session.cmd(msg)
+        if not output.exit_status == 0:
```
Comment (harish-24) [inline io/nvmf/nvmftest.py:131]: same here

File: io/nvmf/nvmftest.py [hunk 3]
```diff
@@ -185,17 +198,21 @@ def test_nvmfdisconnectcfg(self):
...
         """
+        self.session = Session(self.peer_ips[0], user=self.peer_user,
+                               password=self.peer_password)
+        if not self.session.connect():
+            self.fail("failed connecting to peer")
         msg = "nvmetcli clear"
-        cmd = "ssh %s \"%s\"" % (self.peer_ips[0], msg)
-        if process.system(cmd, shell=True, ignore_status=True) != 0:
+        output = self.session.cmd(msg)
+        if not output.exit_status == 0:
```
Comment (harish-24) [inline io/nvmf/nvmftest.py:213]: same here

File: io/nvmf/nvmftest.py [hunk 4]
```diff
@@ -185,17 +198,21 @@ def test_nvmfdisconnectcfg(self):
...
             self.fail("nvmetcli clear config remove on peer")
         msg = "rm -rf /tmp/nvmf.cfg"
-        cmd = "ssh %s \"%s\"" % (self.peer_ips[0], msg)
-        if process.system(cmd, shell=True, ignore_status=True) != 0:
+        output = self.session.cmd(msg)
+        if not output.exit_status == 0:
```
Comment (harish-24) [inline io/nvmf/nvmftest.py:217]: same here as well

**Review by harish-24 — APPROVED** [review]
LGTM

**Pull request merged**